### PR TITLE
[IE][VPU] GatherND layer & tests

### PR DESCRIPTION
**Pull request by maxim-kurin**
Myriad GatherND stage + IE shared tests.  

Related: [Firmware PR](https://github.com/movidius/mdk/pull/14707)

### CODE REVIEW

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:96]: Could you move `gatherND` creation logic to the builders? (lambda & gathernd creation)

**Review by iefode — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (nikita-kud) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:44]: It looks like you need to add "config" to extend the test name to avoid duplication

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (nikita-kud) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:55]: style: `const auto`

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/gather_nd.hpp
Comment (nikita-kud) [inline inference-engine/tests/functional/plugin/shared/include/single_layer_tests/gather_nd.hpp:33]: @iefode is virtual inheritance required? Is it possible that someone would need to inherit from `GatherNDLayerTest`  and `LayerTestsCommon` or its child?
I see that it is often applied. Could you help me figure it out?

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (nikita-kud) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:84]: How about a code like this?

```
std::vector<int> indicesValues(indicesCount * indexSize);
std::generate (indicesValues.begin(), indicesValues.end(), Generator);
```

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/gather_nd.hpp
Comment (iefode) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/gather_nd.hpp:33]: Virtual inheritance was implemented for KMB plugin

Comment (nikita-kud) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/gather_nd.hpp:33]: @iefode thank you for the clarification!

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:84]: Indexes depend  on current position in the tensor.  
If indexSize == 3, indices vector looks like [c, h, w, c, h, w, ...], where c < C, h < H, w < W for tensor with CxHxW size.

Comment (nikita-kud) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:84]: ok, I see

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp]: Done

**Review by nikita-kud — APPROVED** [review]

**Review by iefode — APPROVED** [review]

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:47]: It's better to use `VPU_THROW_UNLESS` here

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:52]: I think we should assert `input precision == output precision`

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:56]: Style: use camelCase

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:80]: Please also provide the name of a layer

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:88]: Is it true that `batch_dims` param always exists?

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp:45]: `_config=` and remove `_` from the target device line

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp]: Done

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:55]: Done

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp]: Done

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp]: Done

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp]: Constructor of GatherND operation has that value = 0 by default. Added it here also.

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp]: Done

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:50]: Please, provide a more informative message. That is with stage type, stage name, the actual number of inputs/outputs and precisions

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:88]: Excess whitespace. Also, provide an actual number of inputs/outputs, please

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:59]: `getOrDefault`?

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:88]: Is there a style guide or `.clang-format` file I could use?

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (maxim-kurin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:66]: As I understand, there always be a value, because in `FrontEnd::parseGatherND`
```
const auto batchDims = layer->GetParamAsInt("batch_dims", 0); // 0 is a default value

stage->attrs().set<int32_t>("batch_dims", batchDims);
```

Comment (Maxim-Doronin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:68]: It is just about further safety. If we implement `addGatherNDStage` then a developer will have the opportunity to set attributes by oneself

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [reply inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp]: > Also, provide an actual number of inputs/outputs, please

I'll add such a rule into the Style guide and send it to you later

File: inference-engine/src/vpu/graph_transformer/include/vpu/stage_builder.hpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/include/vpu/stage_builder.hpp:316]: Where is the implementation? 
If it is not required now, please, remove the declaration

File: inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/stages/gather_nd.cpp:60]:
```
VPU_THROW_UNLESS(inputs()[0]->desc().type() == outputs()[0]->desc().type(),
                 "First input and output must have the same DataType, actual input type is {} and output type is {}", 
                 inputs()[0]->desc().type(), outputs()[0]->desc().type());
```

File: inference-engine/tests/ngraph_functions/src/gather_nd.cpp
Comment (nshchego) [inline inference-engine/tests/ngraph_functions/src/gather_nd.cpp:34]: inputShape[dim + batchDims] ?

File: inference-engine/tests/functional/plugin/myriad/shared_tests_instances/single_layer_tests/gather_nd.cpp
Comment (nshchego) [inline inference-engine/tests/functional/plugin/myriad/shared_tests_instances/single_layer_tests/gather_nd.cpp:50]: Are Layouts parameters really needed?

Comment (iefode) [reply inference-engine/tests/functional/plugin/myriad/shared_tests_instances/single_layer_tests/gather_nd.cpp:50]: This functionality was added by KMB request

**Review by avershov — APPROVED** [review]

**Comment by iefode** [comment]
@nshchego Could you please review again?

**Review by Maxim-Doronin — APPROVED** [review]

**Review by nshchego — APPROVED** [review]

**Comment by maxim-kurin** [comment]
@Maxim-Doronin @nshchego changes after CPU plugin merge

**Review by Maxim-Doronin — APPROVED** [review]

File: inference-engine/tests/ngraph_functions/include/ngraph_functions/builders.hpp
Comment (nshchego) [inline inference-engine/tests/ngraph_functions/include/ngraph_functions/builders.hpp:443]: Is it a duplicate of the function on line #429?

Comment (maxim-kurin) [reply inference-engine/tests/ngraph_functions/include/ngraph_functions/builders.hpp]: good catch. removed.

**Comment by iefode** [comment]
@nshchego Have merged his [PR](#2757) with GatherND tests

**Comment by maxim-kurin** [comment]
> @nshchego Have merged his [PR](#2757) with GatherND tests

Yeah, saw that. I've resolved merge conflicts and now waiting for CI

**Pull request merged**
